src/main/java/org/openstreetmap/atlas/checks/validation/areas/WaterbodyAndIslandSizeCheck.java

```diff
@@ -0,0 +1,328 @@
+                            && !this.isFlagged(member.getEntity().getIdentifier()))
+                    .collect(Collectors.toSet());
+
+            if (!relationMembers.isEmpty())
+            {
+                // Get valid members that fail the surface area size check
+                final Set<RelationMember> invalidPolygonMembers = this
+                        .filterMultiPolygonMembers(relationMembers);
+
+                if (!invalidPolygonMembers.isEmpty())
+                {
+                    // Get instructions for each member
+                    final List<String> instructions = this
+                            .getMemberInstructions(invalidPolygonMembers);
+
+                    if (!instructions.isEmpty())
+                    {
+                        // Create a single flag with each member entity and its given instruction
+                        return Optional.of(instructions).map(instruction -> new CheckFlag(
+                                this.getTaskIdentifier(relation), invalidPolygonMembers.stream()
+                                        .map(RelationMember::getEntity).collect(Collectors.toSet()),
+                                instruction));
+                    }
+                }
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    /**
+     * Helper function that filters multipolygon members.
+     *
+     * @param relationPolygonMembers
+     *            - relation members
+     * @return - a Set of Relation Members
+     */
+    private Set<RelationMember> filterMultiPolygonMembers(
+            final Set<RelationMember> relationPolygonMembers)
+    {
+        return relationPolygonMembers.stream().filter(member ->
+        // Each member must have an inter or outer role
+        (member.getRole().equals(RelationTypeTag.MULTIPOLYGON_ROLE_OUTER) || (member.getRole()
+                .equals(RelationTypeTag.MULTIPOLYGON_ROLE_INNER)
+                // Inner member cannot have natural=rock tag
+                && !Validators.isOfType(member.getEntity(), NaturalTag.class, NaturalTag.ROCK)))
+                // Member surface area be larger or smaller than minimum/maximum constants
+                && this.failsAreaTest(member)).collect(Collectors.toSet());
+    }
+
+    /**
+     * Returns a list of instructions for inner and outer MultiPolygons that fail the size check
+     *
+     * @param relationMembers
+     *            - Members of a Relations
+     * @return List of instructions
+     */
+    private List<String> getMemberInstructions(final Set<RelationMember> relationMembers)
+    {
+
+        final List<String> instructions = new ArrayList<>();
+
+        relationMembers.forEach(member ->
+        {
+            final double areaInMeters = ((Area) member.getEntity()).asPolygon().surface()
+                    .asMeterSquared();
+            final double areaInKilometers = ((Area) member.getEntity()).asPolygon().surface()
```

Comment:
You grab surface area in this method, the `flag` method above, and the `failsAreaTest` method below. Should we abstract this out and stick this logic in a helper?